File: src/main/java/no/nav/fo/forenkletdeploy/service/ApplicationService.java
package no.nav.fo.forenkletdeploy.service;

import no.nav.fo.forenkletdeploy.domain.ApplicationConfig;
import no.nav.fo.forenkletdeploy.provider.TeamProvider;
import no.nav.fo.forenkletdeploy.teams.Team;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import javax.inject.Inject;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class ApplicationService {

    private final TeamProvider teamProvider;
    private static final long MINUTTER = 1000 * 60;
    private Logger logger = LoggerFactory.getLogger(ApplicationService.class.getName());

    @Inject
    public ApplicationService(TeamProvider teamProvider) {
        this.teamProvider = teamProvider;
    }

    @SuppressWarnings("unchecked")
    @Cacheable("applicationlist")
    public List<ApplicationConfig> getApps() {
        return getAllAppConfigurations();
    }

    private List<ApplicationConfig> getAllAppConfigurations() {
        return teamProvider.getTeams().stream()
                .flatMap(team -> team.getApplicationConfigs().stream())
                .collect(Collectors.toList());
    }

    @SuppressWarnings("unchecked")
    @Cacheable("applicationlistbyteam")
    public List<ApplicationConfig> getAppsByTeam(String teamId) {
        return teamProvider.getTeams().stream()
                .filter(team -> team.getId().equals(teamId))
                .flatMap(team -> team.getApplicationConfigs().stream())
                .collect(Collectors.toList());
    }

    @Cacheable("appbyname")
    public ApplicationConfig getAppByName(String name) {
        return getAllAppConfigurations().stream()
                .filter(app -> app.name.equalsIgnoreCase(name))
                .findFirst()
                .get();
    }

    @Scheduled(fixedRate = 5 * MINUTTER)
    public void lastAlleApplicationConfigs() {
        logger.info("Henter og oppdaterer alle applicationConfigs");
        teamProvider.getTeams().forEach(Team::hentApplicationConfigs);
    }

}

File: src/main/kotlin/no/nav/fo/forenkletdeploy/service/ApplicationService.kt
package no.nav.fo.forenkletdeploy.service

import no.nav.fo.forenkletdeploy.ApplicationConfig
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import javax.inject.Inject

@Service
class ApplicationService @Inject
constructor(
    val teamService: TeamService
) {
    val logger: Logger = LoggerFactory.getLogger(ApplicationService::class.java);

    fun getAllApplications() =
            teamService.allTeams.flatMap {
                try {
                    getAppsForTeam(it.id)
                } catch (e: Exception) {
                    logger.error("Kunne ikke hente appConfig for ${it.displayName} via ${it.configUrl}", e)
                    return emptyList<ApplicationConfig>()
                }
            }

    fun getAppsForTeam(teamId: String): List<ApplicationConfig> =
            teamService.getAppsForTeam(teamId)

    fun getAppByName(name: String): ApplicationConfig =
            try {
                getAllApplications().first { it.name.equals(name, ignoreCase = true) }
            } catch (e: NoSuchElementException) {
                throw RuntimeException("Kunne ikke finne '$name' blandt konfigurerte applikasjoner.", e)
            }
}